sklearn/datasets/_arff_parser.py

```diff
@@ -0,0 +1,471 @@
+
+    Returns
+    -------
+    arff_data_new : tuple
+        Subset of arff data with only the include columns indicated by the
+        include_columns argument.
+    """
+    arff_data_new: ArffSparseDataType = (list(), list(), list())
+    reindexed_columns = {
+        column_idx: array_idx for array_idx, column_idx in enumerate(include_columns)
+    }
+    for val, row_idx, col_idx in zip(arff_data[0], arff_data[1], arff_data[2]):
+        if col_idx in include_columns:
+            arff_data_new[0].append(val)
+            arff_data_new[1].append(row_idx)
+            arff_data_new[2].append(reindexed_columns[col_idx])
+    return arff_data_new
+
+
+def _sparse_data_to_array(
+    arff_data: ArffSparseDataType, include_columns: List
+) -> np.ndarray:
+    # turns the sparse data back into an array (can't use toarray() function,
+    # as this does only work on numeric data)
+    num_obs = max(arff_data[1]) + 1
+    y_shape = (num_obs, len(include_columns))
+    reindexed_columns = {
+        column_idx: array_idx for array_idx, column_idx in enumerate(include_columns)
+    }
+    # TODO: improve for efficiency
+    y = np.empty(y_shape, dtype=np.float64)
+    for val, row_idx, col_idx in zip(arff_data[0], arff_data[1], arff_data[2]):
+        if col_idx in include_columns:
+            y[row_idx, reindexed_columns[col_idx]] = val
+    return y
+
+
+def _cast_frame(frame, columns_info):
+    """Cast the columns of a dataframe using the ARFF metadata.
+
+    Parameters
+    ----------
+    frame : dataframe
+        The dataframe to cast.
+
+    columns_info : dict
+        The ARFF metadata for the columns of the dataframe.
+
+    Returns
+    -------
+    frame : dataframe
+        The dataframe with the right casting.
+    """
+    dtypes = {}
+    for name in frame.columns:
+        column_dtype = columns_info[name]["data_type"]
+        if column_dtype.lower() == "integer":
+            dtypes[name] = "Int64"
+        elif column_dtype.lower() == "nominal":
+            dtypes[name] = "category"
+        else:
+            dtypes[name] = frame.dtypes[name]
+    return frame.astype(dtypes)
+
+
+def _post_process_frame(frame, feature_names, target_names):
+    """Post process a dataframe to select the desired columns in `X` and `y`.
+
+    Parameters
+    ----------
+    frame : dataframe
+        The dataframe to split into `X` and `y`.
+
+    feature_names : list of str
+        The list of feature names to populate `X`.
+
+    target_names : list of str
+        The list of target names to populate `y`.
+
+    Returns
+    -------
+    X : dataframe
+        The dataframe containing the features.
+
+    y : {series, dataframe} or None
+        The series or dataframe containing the target.
+    """
+    X = frame[feature_names]
+    if len(target_names) >= 2:
+        y = frame[target_names]
+    elif len(target_names) == 1:
+        y = frame[target_names[0]]
+    else:
+        y = None
+    return X, y
+
+
+def _liac_arff_parser(
```

Comment:
I think this function is mostly a refactor of the current ARFF parse coding. To reduce the size of this PR, can we have an initial PR where we perform this refactor?

Reply:
I can try that.